### ISSUE TITLE
Using hubs inside Hostedservice with kestrel on prod environment

### ISSUE THREAD

**Issue by Tratcher** [aspnet/SignalR#3001]
_From @hrnsky on September 24, 2018 16:46_

Hello all, I don't know where to open this issue weather is an issue or not, hope you will help me. I have a background service, running good, I'm running a project with asp.net core 2.1.4 on ubuntu 18 with nginx 1.14 as proxy with kestrel. Inside this background service, I'm using SignalR hub to send some messages to clients. In this background service, some action happens and pushes the json response to the client with clientID. This scenario runs good while debugging on windows 10 with vs 2017, but doesn't work on prod environment. I'm publishing the project with linux-64 option, then run on ubuntu with systemd service. What should i do ? Is it a good practice to use Hubs inside IHostedService ? Also I'm using this hub inside controllers, nothing going bad. How can i log this situation and how to get the log ? I'm tracking with journalctl, there is no exception and also logging my "hub sent" log message. Thanks

_Copied from original issue: aspnet/Home#3551_

**Comment by anurse** [comment]
@hrnsky you should definitely be able to use `IHubContext` from an `IHostedService`. Can you provide an example of your code and where it's failing? We need a little more detail on what it means when you say it "doesn't work" :).

**Comment by hrnsky** [comment]
@anurse It's good to know that I can use IHubContext inside IHostedService, I think I found the bug, some of not existed routes make application crash somewhere and that makes my connection close, but there is no exception in logs, also no connection closed message. I will update the AllowedHosts section in application.json and fix the routes then try again. Sorry for misunderstood, and thanks for attention.

**Comment by anurse** [comment]
If the application crashes there should definitely be logs. Can you provide a code sample that causes the failure?

**Comment by hrnsky** [comment]
There was no exception but I found this
`Socket closed.
dbug: Microsoft.AspNetCore.SignalR.HubConnectionHandler[6]
OnConnectedAsync ending.
dbug: Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionManager[2]
Removing connection RnssLjImOOlEuHL0B1PD2g from the list of connections.`
I'm really sorry for this, I found the bug that caused the connection close. I thought that was a misconfigured kestrel or SignalR settings. Everything is fine now. IHostedService with Hubs really saved my life, thanks again.
